React I18n: Simplify API

sirreal opened the pull request:
#### Changes proposed in this Pull Request

- Applies suggestion from code review of #39388: use default `en` locale
- Remove `loadLocaleData` prop and simplify provider, let consumers handle context updating
- _Fail fast_ in consumers - if the context is not found, throw an error.

In #39410 it became clear that it's useful to let consumers set and change the `locale` and `localeData` in the context. Mounting the provider with the desired locale after localeData has been fetched is one example. Locale loading is application specific and not a good fit for this early and premature abstraction.

#### Testing instructions

<!-- Demo in #39410 [`/gutenboarding`](https://calypso.live/gutenboarding?branch=gutenboarding/i18n-with-hooks) -->

Demo in [`/gutenboarding`](https://calypso.live/gutenboarding?branch=update/react-i18n-simplification). Initialized with `'es'` locale. You can run `updateLocale( anotherLocaleSlug )`to watch state updates.

The demo includes a `PureComponent` that is a good test that the first render happens with the correct localeData already set.

Closes #39441 (included/superseded)

sirreal committed:
Use "en" for default locale

sirreal committed:
Rename locale loading prop to loadLocaleData

sirreal committed:
react-i18n: Work with localeData

sirreal committed:
react-i18n: Fail fast, name component

jsnajdr commented on packages/react-i18n/src/index.tsx:
```diff
@@ -65,6 +55,9 @@ function makeI18n( i18nLocale: string ) {
 	const i18n = React.useContext( I18nContext );
+	if ( i18n === undefined ) {
+		throw new Error( 'useI18n must be descendent of <I18nProvider />' );
+	}
```
`React.createContext()` takes the default context value as a parameter. And this value will be used by the consumers if no provider is mounted.

In other words, using the consumer without a provider is a valid and well-defined state. It will use the default.
We don't need to detect it as error state.

We'll appreciate this feature especially in unit tests, where we won't need to bother mounting the `I18nProvider` if we're not testing i18n specifically.

sirreal replied:
I've addressed this in 8751ed04d5b0199c8947f6d627688041753cca76, please see what you think.

Some module state tracks the last locale, and a function is invoked on every render. This invokes `setLocaleData` when `locale` has changed. This is unappealing, but does match the singleton paradigm currently used in `@wordpress/i18n`.

jsnajdr commented on packages/react-i18n/src/index.tsx:
```diff
@@ -88,15 +82,27 @@ type Optionalize< T extends K, K > = Omit< T, keyof K >;
+				}
+				return (
+					<InnerComponent
+						{ ...i18n }
```
Why did you decide to abandon the `useI18n` hook here? It's a common practice these days to implement HOC wrappers using hooks.

sirreal replied:
This was mostly related to the other point. I noticed `withI18n` wasn't working well with a `PureComponent` and tried a few different things. I've gone back to the `useI18n` form.

jsnajdr commented on packages/react-i18n/src/index.tsx:
```diff
@@ -88,15 +82,27 @@ type Optionalize< T extends K, K > = Omit< T, keyof K >;
 	);
+	WrappedComponent.displayName = `withI18n( ${ WrappedComponent.displayName ||
+		WrappedComponent.name ||
+		'Component' } )`;
```
Can you use `createHigherOrderComponent` from `@wordpress/compose`? Or is that not TypeScript friendly?

sirreal replied:
Done, the types work nicely 👍

sirreal committed:
DROPME: Debug with PureComponent

sirreal committed:
Drop error on no context

sirreal committed:
Use createHigherOrderComponent with useI18n

sirreal committed:
Single pass global setLocaleData

sirreal committed:
Add DOM to libs

sirreal committed:
Pass I18nProps to createHightOrderCompontnt type

sirreal commented on packages/react-i18n/tsconfig.json:
```diff
@@ -1,7 +1,7 @@
 	"compilerOptions": {
 		"target": "ES5",
-		"lib": [ "ES2015" ],
+		"lib": [ "DOM", "ES2015" ],
```
`DOM` lib is required for `@wordpress/compose`.

jsnajdr replied:
Is that an issue with the `@types/wordpress__compose` package? The `createHigherOrderComponent` function itself doesn't depend on DOM. But `@wordpress/compose` is a somewhat random bag of various utilities, where some of them use DOM (e.g., `withGlobalEvents`). Then importing the whole `@types/wordpress__compose/index.d.ts` depends on DOM, too.

I think this is worth reporting as a Gutenberg issue. As @sgomes says, utility modules that bundle random things together can cause problems. Not only with tree shaking, but now also with types 😢

sirreal replied:
I believe it was the `WindowEventMap` type [used here](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/d129b9b777901659b1559f9355f9548496e72a2b/types/wordpress__compose/higher-order/with-global-events/index.d.ts#L4).

The type declarations do seem to have a nice tree-shakeable structure, however it does seem that types are eagerly evaluated.

> I think this is worth reporting as a Gutenberg issue.

It may be. However, type declarations are strictly third part ay this time. Gutenberg does seem open to generating type declarations, but this sort of issue seems to be somewhere between Gutenberg proper and the type declarations.

sirreal commented:
I added 0b174827926b8cbc8cdcd16ef50efa52185d6287 which adds easier testing/debugging to this PR but is not intended to be included.

Thanks for the valuable feedback, @jsnajdr. I've addressed your points and I'd appreciate if you have another look 🙇

jsnajdr commented on client/landing/gutenboarding/components/header/index.tsx:
```diff
@@ -179,3 +178,18 @@ const Header: FunctionComponent< Props > = ( { prev } ) => {
+const BackButton = withI18n(
+	class InnerBackButton extends React.PureComponent< I18nProps & { prev?: string } > {
+		render() {
+			const { prev, __ } = this.props;
```
Did you mention you had some issues getting the i18n updates through the `PureComponent` barrier 🙂 

One reason for that is that `__` is always the same object, no matter if the underlying `i18n` object got updated or not.

Usually, the changed `i18nLocale` prop should be enough to force rerender of a `PureComponent`, but sometimes (especially in Calypso), the translations can get updated with `i18nLocale` unchanged. E.g., in the community translator.

A good technique to solve that is to not return the raw `__` function from `makeI18n`, but `bind` it to create a new function with unique identity.

https://github.com/Automattic/wp-calypso/blob/master/packages/i18n-calypso/src/use-translate.js#L8-L10

sirreal replied:
The reason I had difficulty with that was caused by the issue you noted in your review here:

https://github.com/Automattic/wp-calypso/pull/39480#discussion_r380139496

As long as locale data is set before the first render, `PureComponent`s are happy. Subsequent updates to the `i18nLocale` prop trigger the desired rerenders. That was my motivation for adding the `i18nLocale` prop.

> A good technique to solve that is to not return the raw `__` function from `makeI18n`, but `bind` it to create a new function with unique identity.

I considered this, but preferred relying on `i18nLocale` changes. The issue should be fixed by the changes in this PR.

yansern commented:
This works perfectly fine. Accessing `/gutenboarding/es` shows the spanish language on the header.

yansern reviewed (approved):
This works perfectly fine. Accessing `/gutenboarding/es` shows the Spanish language on the header.

sirreal commented:
I plan to land this when the tests pass. @jsnajdr if you have any additional feedback, again I'll be happy to address it 👍 

Thanks everyone.

sirreal committed:
Revert "DROPME: Debug with PureComponent"

This reverts commit 0b174827926b8cbc8cdcd16ef50efa52185d6287.

sirreal committed:
Update package-lock with wp-compose

sirreal commented:
Rebased and reverted the demonstration commits.

jsnajdr commented on packages/react-i18n/src/index.tsx:
```diff
@@ -21,35 +32,25 @@ interface Props {
-	onLocaleChange = () => Promise.resolve( {} ),
+	localeData = {},
 } ) => {
+	updateLocaleOnChange( locale, localeData );
```
The render function should be pure, without side effects. But `updateLocaleOnChange` is updating a global variable.

And the `lastLocale` variable is shared by all components that use `react-i18n`. It creates another obstacle on the way to working SSR, because different React trees rendered in different locales in parallel (responding to parallel HTTP requests) will fight over this variable.

I think that the `useI18n` hook and context shouldn't call `setLocaleData()` at all. In Gutenboarding for example, let's make it part of the code that figures out the locale and fetches the translation file. `setLocaleData()` belongs there naturally, IMO:
```js
const locale = await figureOutUserLocale();
const localeData = await fetchTranslationFile( locale );
i18n.setLocaleData( localeData );

ReactDOM.render(
  <I18nProvider locale={ locale } />
    <App />
  </I18nProvider>
);
```

sirreal replied:
For context: I'm expecting most of this to be thrown away when https://github.com/WordPress/gutenberg/issues/19210 is resolved and I don't really believe this package provides much value until it can support the features you've mentioned.

I'll make the changes you suggested.

jsnajdr replied:
I agree: when we have multiple instances available, we'll simply create a new `i18n` instance for every locale, load the data into it, and pass it as the context value. Is that what you had in mind?

sirreal replied:
Remove locale setting from 7a3eef9fe91370833eabed77a28b713ed784cf30 which makes the package very simple.

sirreal replied:
> I agree: when we have multiple instances available, we'll simply create a new `i18n` instance for every locale, load the data into it, and pass it as the context value. Is that what you had in mind?

Yes, that's the kind of detail I was expecting to handle in this library. I was probably getting ahead of myself with the current functionality of `@wordpress/i18n` 😉

sirreal committed:
Remove locale data setting

jsnajdr reviewed (approved):
This is still not the best and final implementation, but a good step forward nevertheless :+1:
